crates/milli/src/search/new/geo_sort.rs

```diff
@@ -267,24 +265,79 @@ impl<'ctx, Q: RankingRuleQueryTrait> RankingRule<'ctx, Q> for GeoSort<Q> {
+                if geo_candidates.contains(id) {
+                    let distance = distance_between_two_points(&self.point, &point).round() as u32;
+                    if let Some((point0, bucket_distance)) = current_distance.as_ref() {
+                        if bucket_distance != &distance {
+                            // different distance, point belongs to next bucket
+                            put_back(&mut self.cached_sorted_docids, (id, point));
+                            return Ok(Some(RankingRuleOutput {
+                                query,
+                                candidates: current_bucket,
+                                score: ScoreDetails::GeoSort(score_details::GeoSort {
+                                    target_point: self.point,
+                                    ascending: self.ascending,
+                                    value: Some(point0.to_owned()),
+                                }),
+                            }));
+                        } else {
+                            // same distance, point belongs to current bucket
+                            current_bucket.push(id);
+                            // remove from cadidates to prevent it from being added to the cache again
+                            geo_candidates.remove(id);
+                        }
+                    } else {
+                        // first doc in current bucket
+                        current_distance = Some((point, distance));
+                        current_bucket.push(id);
+                        geo_candidates.remove(id);
+                    }
+                }
+            } else {
+                // cache exhausted, we need to refill it
+                self.fill_buffer(ctx, &geo_candidates)?;
+
+                if self.cached_sorted_docids.is_empty() {
+                    // candidates exhausted, exit
+                    if let Some((point0, _)) = current_distance.as_ref() {
+                        return Ok(Some(RankingRuleOutput {
+                            query,
+                            candidates: current_bucket,
+                            score: ScoreDetails::GeoSort(score_details::GeoSort {
+                                target_point: self.point,
+                                ascending: self.ascending,
+                                value: Some(point0.to_owned()),
+                            }),
+                        }));
+                    } else {
+                        return Ok(Some(RankingRuleOutput {
+                            query,
+                            candidates: universe.clone(),
+                            score: ScoreDetails::GeoSort(score_details::GeoSort {
+                                target_point: self.point,
+                                ascending: self.ascending,
+                                value: None,
+                            }),
+                        }));
```

Comment:
Hey, I don’t understand why we’re returning nothing here?
Shouldn’t we continue with the loop and check if we have other candidates matching our bucket?

It would be nice if you could add a test where two candidates must be grouped together but are not part of the same bucket, as this is now a bit tricky to handle.

Reply:
If self.cached_sorted_docids is still empty after refill, it means we have exhausted the entire candidate set. At this time, if current_bucket is not empty, I should return current_bucket. But if current_bucket is empty, I can only return an empty result set.

Also, when you say "grouped together", do you mean they appear in the cache at the same time?

Reply:
Aaaah yes sorry my bad, I thought the else was applied to the other branch my bad 🤦 
So you’re actually doing what I was expecting 👌